WPDB driver for Doctrine [MAILPOET-6142]

JanJakes opened the pull request:
## Description

This PR implements WPDB driver for Doctrine and replaces the current PDO with this new driver.

**It includes:**
1. Removing old models and Paris, Idiorm and Sudzy.
2. Replacing all PDO-specific expressions.
3. Using single quotes for strings in SQL queries (more standard and required by WPDB).
4. WPDB driver, connection, statement, result, and parameter handling.
6. Exception handling with SQL error codes and SQLSTATE for conversion to their respective DBAL exceptions.
7. Removing dependency on PDO altogether.
8. Better cleanup of unused Doctrine drivers.

**It doesn't include:**
1. WPDB cache integration. This is something we could consider in the future, but as we're not using any caching adapters in Doctrine at the moment, it makes sense to first replace implement the WPDB driver without additional caching.
2. SQLite-specific code. The WP Playground uses SQLite, but I'm not sure if any SQLite specifics are needed (such as configuring SQLite Platform for Doctrine), as the WP Playground may be emulating some MySQL-like behavior, which could even mean any SQLite-specific settings could be counter-productive.

## Code review notes

There are some fixes to the existing Doctrine code (like strings to single quotes and passing parameters). If you find a way to check I didn't miss any more of those, it would be great. I tried some full-text and pattern search, but a second pair of eyes is always useful.

## QA notes

We need a general plugin smoke check, trying to trigger as much functionality as possible. This replaces the database driver, so if any issues appear, it could be anywhere. That said, our test being green already gives us some confidence.

There was a significant change in the underlying code for working with times and time zones. Please check that the times we present to users (listings, logs, help page) are correct and that scheduling newsletters and post notifications works as expected. We store time in the DB in GMT, but we always present time in the site's timezone, and we also expect the user to enter time in the site's timezone. 

## Linked PRs

https://github.com/mailpoet/mailpoet-premium/pull/894

## Linked tickets

[MAILPOET-6142]

## After-merge notes

_N/A_

## Tasks

- [x] I followed [best practices](https://codex.wordpress.org/I18n_for_WordPress_Developers) for translations
- [x] I added sufficient test coverage
- [x] I embraced TypeScript by either creating new files in TypeScript or converting existing JavaScript files when making changes


[MAILPOET-6142]: https://mailpoet.atlassian.net/browse/MAILPOET-6142?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

JanJakes committed:
Replace old model table name constants with Doctrine repositories

[MAILPOET-6142]

JanJakes committed:
Remove all old models and their setup

[MAILPOET-6142]

JanJakes committed:
Remove Paris, Idiorm, and Sudzy

[MAILPOET-6142]

JanJakes committed:
Replace PDO types with Doctrine's ParameterType

[MAILPOET-6142]

JanJakes committed:
Replace deprecated array types with Doctrine's ArrayParameterType

[MAILPOET-6142]

JanJakes committed:
Implement basic WPDB Doctrine driver with connection, statement, and result stubs

[MAILPOET-6142]

JanJakes committed:
Implement Connection::exec() and other simple connection methods

[MAILPOET-6142]

JanJakes committed:
Implement Result and Connection::query()

[MAILPOET-6142]

JanJakes committed:
Use single quotes for strings in SQL queries

This is more standard SQL and required by WPDB.

[MAILPOET-6142]

JanJakes committed:
Implement Statement and Connection::prepare()

[MAILPOET-6142]

JanJakes committed:
Implement parameter conversion and named parameters

[MAILPOET-6142]

JanJakes committed:
Replace deprecated Statement::bindParam() and Statement::exec() with params

[MAILPOET-6142]

JanJakes committed:
Split mutli-query into multiple statements

WPDB doesn't support multi queries.

[MAILPOET-6142]

JanJakes committed:
Use mysqli to run test cleanups

[MAILPOET-6142]

JanJakes committed:
Use WPDB driver for Doctrine

[MAILPOET-6142]

JanJakes committed:
Remove PDO

[MAILPOET-6142]

JanJakes committed:
Fix last insert ID type check

[MAILPOET-6142]

JanJakes committed:
Fix missing array key

[MAILPOET-6142]

JanJakes committed:
Pass SQLSTATE and code to query exceptions so they are converted to DBAL exceptions properly

The conversion is done by MailPoetVendor\Doctrine\DBAL\Driver\API\MySQL\ExceptionConverter
that is returned from MailPoetVendor\Doctrine\DBAL\Driver\AbstractMySQLDriver.

[MAILPOET-6142]

JanJakes committed:
Cleanup unused Doctrine drivers and other files

[MAILPOET-6142]

costasovo commented on mailpoet/lib/Config/Database.php:
```diff
@@ -1,53 +0,0 @@
-      define('MP_STATS_NOTIFICATIONS_TABLE', Env::$dbPrefix . 'stats_notifications');
-      define('MP_USER_FLAGS_TABLE', Env::$dbPrefix . 'user_flags');
-      define('MP_FEATURE_FLAGS_TABLE', Env::$dbPrefix . 'feature_flags');
-      define('MP_DYNAMIC_SEGMENTS_FILTERS_TABLE', Env::$dbPrefix . 'dynamic_segment_filters');
```
I have a concern that some plugins that integrate with MailPoet may use these constants in raw SQL queries. So maybe we could keep the definitions and perhaps mark them as deprecated (if that's possible)

costasovo replied:
I added a helper class which reintroduces the constants.

costasovo commented on mailpoet/lib/Doctrine/WPDB/Result.php:
```diff
@@ -4,40 +4,69 @@
+
   public function fetchNumeric() {
-    // TODO: Implement fetchNumeric() method.
+    $value = $this->result[$this->cursor++] ?? null;
```
Just a suggestion. If I understand correctly, instead of the `$value` variable, we could use `$row` because `$result` is an array of rows, and each row is an associative array, right?
So, I would suggest renaming `$value` to `$row` in the whole file. For example later in `fetchOne` is see `reset($value)` but `reset($row)` reads much better for me and it is easier to understand that it returns the value from first column.

costasovo commented on mailpoet/lib/Doctrine/WPDB/ConvertParameters.php:
```diff
@@ -0,0 +1,84 @@
+    [, $value, $type] = $this->params[$key];
+
+    // WPDB doesn't support NULL values. We need to handle them explicitly.
+    if ($value === null) {
+      $this->buffer[] = 'NULL';
+      return;
+    }
```
This replicates the current behavior, but I see in the test it generates queries with conditions `WHERE column = null`, which is always false. I wonder how many such cases we have. I guess we can try to log this in our tests.

JanJakes replied:
@costasovo Did you manage to figure out how many of such cases we have in our codebase? That said, I think we'll need to keep this behavior to satisfy Doctrine's behavior. What I mean is that if the ORM can generate such queries, we should probably support them on the DBAL level as well.

costasovo replied:
@JanJakes I ran the tests and logged all cases of null arguments that appeared after WHERE or JOIN. In the end, there weren't many places (actually 5), but they were called repeatedly, so the log looked quite big. 

Here is the code I used to check which cases to log to skip null in values:
```php
$shouldLog = false;
foreach ($this->buffer as $bufferPart) {
  if ((stripos($bufferPart, 'WHERE') !== false) || (stripos($bufferPart, 'JOIN') !== false)) {
    $shouldLog = true;
    break;
  }
}
```

I added two commits to address a few cases. Most of the logged occurrences came from queries created by high-level "findBy" methods, but Doctrine seems to be smart enough to cover those.
I found that, for example:
```php
$task = $this->scheduledTasksRepository->findOneBy(
      [
        'status' => [ScheduledTaskEntity::STATUS_SCHEDULED, null],
      ]
    );
```
is converted to
 ```sql
 (status IN ('scheduled', null) OR status IS NULL)
 ```
 
 I went through rest of the cases and added two commits, but they were more like nice to have.
 
 So I think we are fine.

costasovo reviewed (commented):
@JanJakes This is awesome! 🎉 I'm super excited to see this is happening!

I added a few comments. As you are on the rotation I'll look into this tomorrow.

costasovo committed:
Reintroduce legacy DB table name constants for backward compatibility

[MAILPOET-6142]

costasovo committed:
Rename $value to $row in WPDB/Result for better readability

[MAILPOET-6142]

costasovo committed:
Improve SubscriberLifetieEmailCountTest to use saved entity

The test was using an empty entity which was not presisted. This may
potentially cause issues as in some places, an entity with a proper type might be expected.
It was already causing that an SQL query ran with WHERE type = null condition which is always false.
[MAILPOET-6142]

costasovo committed:
Skip null ids when deleting WP list subscribers with invalid emails

Theoretically, we may end up running the delete query with a condition like
WHERE wp_user_id IN (null, null);
This is a tiny optimization to skip null IDs to avoid running such a query.
[MAILPOET-6142]

Aschepikov commented:
Hi @JanJakes there is some failed test, can you check it pls

costasovo committed:
Use UTC time in TimestampListener

$this->wp->currentTime('timestamp') adds 'gmt_offset' value.
This was needed with PDO connection because the connection
used @session.time_zone set to the same offset value.

[MAILPOET-6142]

costasovo committed:
Improve LastSubsribedAtListener to use fresh now

The cached now may cause the value to not be exact.
[MAILPOET-6142]

costasovo committed:
Replace usage of WP's current_time without gmt parameter in libs

This commit replaces usages by Carbon::now or in case we need a timestamp it
keeps current_time but adds the gtm parameter as true.
[MAILPOET-6142]

costasovo committed:
Replace current_time usage in tests

[MAILPOET-6142]

costasovo committed:
Pass GMT offset from WordPress settings to JS

[MAILPOET-6142]

costasovo committed:
Add formatFromGmt etc to MailPoet.Date

This set of function expects input in GMT and add site's offset
They are meant to be used for data coming from server in GMT
[MAILPOET-6142]

costasovo committed:
Replace simple usages of date formating with offet aware methods

[MAILPOET-6142]

costasovo committed:
Use proper time in scheduling calendar on send page

We add offset before we display time and we remove when we pass it to callback.
[MAILPOET-6142]

costasovo committed:
Show proper times for action scheduler info on help page

[MAILPOET-6142]

costasovo committed:
Use isInFutureGmt to check future on send page

[MAILPOET-6142]

costasovo committed:
Show proper time in listing for scheduled newsletter

[MAILPOET-6142]

costasovo committed:
Fix times in task lists on Help > System Status

The time adjustments are not necessary because we rely on WP setting
timezone to UTC.
Manipulation in adjustTimezone difference was causing us to display GMT time.
[MAILPOET-6142]

costasovo committed:
Fix values in list of times in scheduling form

[MAILPOET-6142]

costasovo committed:
Fix scheduling post notifications

The user picks time in a select box, and we expect they pick a time
in the site's timezone so the saved cron expression is in the site's timezone.
When we calculate the next run date, we do it in the site's timezone and
to get UTC time, we convert it.
[MAILPOET-6142]

costasovo committed:
Set GMT timezone to WPDB

[MAILPOET-6142]

costasovo committed:
Remove adjustForTimezoneDifference functionality

This is no longer used. It was used in just a couple of places
and was fixing a difference between PHP time zone and browser time.
We rely on WordPress to set PHP timezone to UTC and we use the site's
timezone on the client so the fix is no longer useful.
[MAILPOET-6142]

costasovo committed:
Use DBAL Driver middleware to handle the correct connection timezone

I looked into using DBAL events for this but found that DBAL events
are deprecated so I went with middleware as they are recommended
as a replacement for the DBAL events.
[MAILPOET-6142]

costasovo committed:
Remove unused date.ts methods and rename the prefixed methods

There is no use case for the methods without the translation to site's timezone.
When the input doesn't specify the timezone we treat it as utc.
When the input provides timezone (e.g. '2024-08-13 10:00+02:00' or new Date('2024-08-13 10:00+02:00'))
We respect the timezone of the input and print the result in the site's timezone.

When the site's timezone is set +01:00

'2024-08-13 10:00' prints as 2024-08-13 11:00
'2024-08-13 10:00:+02:00' prints as 2024-08-13 09:00
new Date('2024-08-13 10:00:+02:00') prints as 2024-08-13 09:00
new Date('2024-08-13 10:00:+00:00') prints as 2024-08-13 11:00

[MAILPOET-6142]

costasovo committed:
Replace usages of prefixed date methods with renamed variants

[MAILPOET-6142]

costasovo committed:
Get rid of unnecessary usages of .toISOString

[MAILPOET-6142]

costasovo committed:
Rename methods for manipulating date in scheduling UI

[MAILPOET-6142]

costasovo committed:
Get rid of unnecessary parameter in the Scheduler

The fromTimestamp parameter was used only in tests.
This commit replaces the parameter in tests by mocks and removes it.
 [MAILPOET-6142]

costasovo committed:
Improve readability of scheduler code

[MAILPOET-6142]

costasovo committed:
Add timezone info to system report

 [MAILPOET-6142]

Aschepikov commented:
Hi @JanJakes there are some conflicts can you check it pls

Aschepikov reviewed (approved):
GJ!
